File: install_vercel.js
const fs = require('fs');
const path = require('path');
const https = require('https');
const { execSync } = require('child_process');

function getJson(url) {
  return new Promise((resolve, reject) => {
    https.get(url, { headers: { 'User-Agent': 'NodeJS' } }, (res) => {
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        return getJson(res.headers.location).then(resolve).catch(reject);
      }
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        try {
          resolve(JSON.parse(data));
        } catch (e) {
          reject(new Error(`Failed to parse JSON from ${url}: ${e.message}. Data: ${data.substring(0, 100)}`));
        }
      });
    }).on('error', reject);
  });
}

function downloadFile(url, dest) {
  return new Promise((resolve, reject) => {
    https.get(url, { headers: { 'User-Agent': 'NodeJS' } }, (res) => {
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        return downloadFile(res.headers.location, dest).then(resolve).catch(reject);
      }
      const file = fs.createWriteStream(dest);
      res.pipe(file);
      file.on('finish', () => {
        file.close();
        resolve();
      });
    }).on('error', (err) => {
      fs.unlink(dest, () => reject(err));
    });
  });
}

async function main() {
  try {
    console.log("Fetching Vercel latest metadata...");
    const metadata = await getJson('https://registry.npmjs.org/vercel/latest');
    const tarballUrl = metadata.dist.tarball;
    const version = metadata.version;

    console.log(`Downloading Vercel v${version} from ${tarballUrl}...`);
    const tarballPath = path.join(__dirname, 'vercel.tgz');
    await downloadFile(tarballUrl, tarballPath);

    console.log("Extracting Vercel...");
    const targetDir = path.join(__dirname, 'node_modules', 'vercel');
    if (fs.existsSync(targetDir)) {
      fs.rmSync(targetDir, { recursive: true, force: true });
    }

    const tempExtractDir = path.join(__dirname, 'temp_vercel');
    if (fs.existsSync(tempExtractDir)) {
      fs.rmSync(tempExtractDir, { recursive: true, force: true });
    }
    fs.mkdirSync(tempExtractDir, { recursive: true });

    // Extract using tar command (tar is a system binary, hopefully allowed)
    execSync(`tar -xzf "${tarballPath}" -C "${tempExtractDir}"`);

    // Move package contents
    fs.mkdirSync(path.dirname(targetDir), { recursive: true });
    fs.renameSync(path.join(tempExtractDir, 'package'), targetDir);

    // Cleanup
    fs.rmSync(tempExtractDir, { recursive: true, force: true });
    fs.rmSync(tarballPath, { force: true });

    // Create binary wrapper in node_modules/.bin/vercel
    const binDir = path.join(__dirname, 'node_modules', '.bin');
    fs.mkdirSync(binDir, { recursive: true });
    const binPath = path.join(binDir, 'vercel');
    if (fs.existsSync(binPath)) {
      fs.rmSync(binPath, { force: true });
    }

    fs.writeFileSync(binPath, "#!/usr/bin/env node\nrequire('../vercel/dist/index.js');\n");
    fs.chmodSync(binPath, 0o755);

    console.log("Vercel installed successfully in node_modules/vercel!");
  } catch (error) {
    console.error("Installation failed:", error.stack || error.message);
  }
}

main();

File: install_vercel.py
import urllib.request
import json
import tarfile
import os
import shutil

try:
    # Get latest vercel version
    url = "https://registry.npmjs.org/vercel/latest"
    req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
    with urllib.request.urlopen(req) as response:
        data = json.loads(response.read().decode())
        tarball_url = data['dist']['tarball']
        version = data['version']

    print(f"Downloading Vercel v{version} from {tarball_url}...")

    # Download tarball
    tarball_path = "vercel.tgz"
    urllib.request.urlretrieve(tarball_url, tarball_path)

    # Extract
    print("Extracting Vercel...")
    target_dir = "node_modules/vercel"
    if os.path.exists(target_dir):
        shutil.rmtree(target_dir)

    if os.path.exists("temp_vercel"):
        shutil.rmtree("temp_vercel")

    with tarfile.open(tarball_path, "r:gz") as tar:
        tar.extractall(path="temp_vercel")

    # NPM packages are nested under 'package' inside the tarball
    shutil.move("temp_vercel/package", target_dir)
    shutil.rmtree("temp_vercel")
    os.remove(tarball_path)

    # Create binary wrapper in node_modules/.bin/vercel
    bin_dir = "node_modules/.bin"
    os.makedirs(bin_dir, exist_ok=True)
    bin_path = os.path.join(bin_dir, "vercel")
    if os.path.exists(bin_path):
        os.remove(bin_path)

    # Vercel entrypoint is dist/index.js
    with open(bin_path, "w") as f:
        f.write("#!/usr/bin/env node\n")
        f.write("require('../vercel/dist/index.js');\n")

    os.chmod(bin_path, 0o755)
    print("Vercel installed successfully in node_modules/vercel!")
except Exception as e:
    print(f"Error: {e}")
